### PR TITLE
JAX zero copy

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
This PR omits the copy when the new executor is used with JAX >= 0.4.16.
When JAX is older or the pipeline doesn't use the dynamic_executor, the copy is still performed.

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-4117

### CODE REVIEW

**Comment by mzient** [comment]
!build

File: dali/python/nvidia/dali/plugin/jax/clu.py [hunk 1]
```diff
@@ -82,7 +82,7 @@ class DALIGenericPeekableIterator(DALIGenericIterator):
                 is called internally automatically.
     last_batch_policy: optional, default = LastBatchPolicy.FILL
                 What to do with the last batch when there are not enough samples in the epoch
-                to fully fill it. See :meth:`nvidia.dali.plugin.base_iterator.LastBatchPolicy`
+                to fully fill it. See :meth:`nvidia.dali.plugin.base_iterator.LastBatchPolicy`.
```
Comment (mzient) [inline dali/python/nvidia/dali/plugin/jax/clu.py:85]: There's a test that compares docstrings and it failed before. Apparently it's not run in CI (?).

File: dali/python/nvidia/dali/plugin/jax/integration.py [hunk 2]
```diff
@@ -16,31 +16,46 @@
...
+        return jax_array.device
 
-    _jax_version_pre_0_4_16 = Version(jax.__version__) < Version("0.4.16")
-    return _jax_version_pre_0_4_16
+elif Version(jax.__version__) >= Version("0.4.27"):
 
+    def _jax_device(jax_array):
+        devs = jax_array.devices()
+        if len(devs) != 1:
+            raise RuntimeError("The must be associated with exactly one device")
```
Comment (szkarpinski) [inline dali/python/nvidia/dali/plugin/jax/integration.py:35]: The what?

**Review by szkarpinski — APPROVED** [review]

File: dali/python/nvidia/dali/plugin/jax/integration.py [hunk 3]
```diff
@@ -16,31 +16,46 @@
...
 
-def _jax_has_old_dlpack():
-    global _jax_version_pre_0_4_16
-    if _jax_version_pre_0_4_16 is not None:
-        return _jax_version_pre_0_4_16
+if Version(jax.__version__) >= Version("0.4.31"):
 
-    from packaging.version import Version
+    def _jax_device(jax_array):
+        return jax_array.device
```
Comment (stiepan) [inline dali/python/nvidia/dali/plugin/jax/integration.py:28]: I couldn't find the documentation for old JAX releases, but looking at the code, the method version used to raise an error for multi-dev array, while this property (according to the docs) returns sharding. So this call works fine with the multi-dev arrays while the other two variants don't.

**Review by stiepan — APPROVED** [review]

**Pull request merged**